### PR TITLE
WIP: Add require signed commits feature

### LINKED ISSUE #4249: Allow signed commits to be required
GitHub allows protected branches to require signed commits. This is a blocking issue for adoption of Gitea by some projects.

### PR DESCRIPTION
**Pull request by lunny**
Fix #4249, This PR is not work yet because on prereceive hook, you should not get the commit information from the repository.

### CODE REVIEW

File: routers/private/hook.go [hunk 1]
```diff
@@ -119,6 +120,48 @@ func HookPreReceive(ctx *macaron.Context) {
...
+			gitRepo, err := git.OpenRepository(repo.RepoPath())
+			if err != nil {
+				log.Error("Unable to find the git repository %s/%s: %v", ownerName, repoName, err)
+				ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+					"err": fmt.Sprintf("Unable to find the git repository %s/%s", ownerName, repoName),
+				})
+				return
+			}
+
+			stdout, err := git.NewCommand("rev-list", oldCommitID+"..."+newCommitID).RunInDirBytes(repo.RepoPath())
```
Comment (zeripath) [inline routers/private/hook.go:135]: In pre-receive the new commits and objects are not placed directly into the object tree (depending on your git version of course). They are placed in to Quarantine and/or various other places.

You need to set the environment for any git commands to tell it to look in those directories for objects too.

https://github.com/go-gitea/gitea/blob/0a004a69cdcad2fa13d3d314cb984718c88a2940/routers/private/hook.go#L67-L82

I wouldn't use OpenRepository either just use the git repo commands directly. If you want to use gogit commands I'm fairly certain you can set its environs similarly.

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1336,6 +1336,8 @@ settings.protect_merge_whitelist_teams = Whitelisted teams for merging:
 settings.protect_check_status_contexts = Enable Status Check
 settings.protect_check_status_contexts_desc = Require status checks to pass before merging Choose which status checks must pass before branches can be merged into a branch that matches this rule. When enabled, commits must first be pushed to another branch, then merged or pushed directly to a branch that matches this rule after status checks have passed. If no contexts are selected, the last commit must be successful regardless of context.
 settings.protect_check_status_contexts_list = Status checks found in the last week for this repository
+settings.requrie_signed_commits = Require signed commits
```
Comment (zeripath) [inline options/locale/locale_en-US.ini:1339]:
```suggestion
settings.require_signed_commits = Require signed commits
```

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -1336,6 +1336,8 @@ settings.protect_merge_whitelist_teams = Whitelisted teams for merging:
...
+settings.requrie_signed_commits_desc = Commits pushed to this branch must have verified signatures.
```
Comment (zeripath) [inline options/locale/locale_en-US.ini:1340]:
```suggestion
settings.require_signed_commits_desc = Commits pushed to this branch must have verified signatures.
```

File: templates/repo/settings/protected_branch.tmpl [hunk 4]
```diff
@@ -136,6 +136,14 @@
 						</div>
 					</div>
 
+					<div class="field">
+						<div class="ui checkbox">
+							<input class="require-signedcommits" name="require_signed_commits" type="checkbox" {{if .Branch.RequireSignedCommits}}checked{{end}}>
+							<label>{{.i18n.Tr "repo.settings.requrie_signed_commits"}}</label>
```
Comment (zeripath) [inline templates/repo/settings/protected_branch.tmpl:142]:
```suggestion
							<label>{{.i18n.Tr "repo.settings.require_signed_commits"}}</label>
```

File: templates/repo/settings/protected_branch.tmpl [hunk 5]
```diff
@@ -136,6 +136,14 @@
...
+							<p class="help">{{.i18n.Tr "repo.settings.requrie_signed_commits_desc"}}</p>
```
Comment (zeripath) [inline templates/repo/settings/protected_branch.tmpl:143]:
```suggestion
							<p class="help">{{.i18n.Tr "repo.settings.require_signed_commits_desc"}}</p>
```

**Review by zeripath — COMMENTED** [review]
Some spelling mistakes and a hint as to how to get the git commands to work

**Comment by zeripath** [comment]
Of course, this doesn't solve the issue of merge and web commits - we still need some way of indicating whether such a commit would be signed and if not possibly disable their merge/submit buttons.

As an aside what happens when you try to use the web editor to commit to a protected branch? Presumably it would fail - but do we nicely disable the submit button?

**Comment by guillep2k** [comment]
> 
> 
> Of course, this doesn't solve the issue of merge and web commits - we still need some way of indicating whether such a commit would be signed and if not possibly disable their merge/submit buttons.
> 
> As an aside what happens when you try to use the web editor to commit to a protected branch? Presumably it would fail - but do we nicely disable the submit button?

If I understand your question correctly, yes: the only option available when the branch is protected is "create new branch".

**Comment by zeripath** [comment]
@guilep2k excellent. I'm sure I've looked at this code before but just forgot.

Ok. So for CRUD we'd need to adjust whatever code does that check to add a check using:

https://github.com/go-gitea/gitea/blob/5a62ae5cbf47cf6537f6c95aba4b1d04dea5f5af/models/repo_sign.go#L169

`tmpBasePath` just has to be either the repo or a temporary copy. You don't have to have a temporary copy for this one. If that test fails then we would have to determine why or just say that the branch requires signed commits. It could be that that method could be changed to return the proximal failure reason.

For merges it could just be on attempt to merge that it fails immediately, we do an individual test on display of the UI if the merge button would be displayed or we could cache the results of the individual tests and recombine them on display at the UI depending on the users 2fa status if that matters.

**Comment by lunny** [comment]
replaced by #9708

**Pull request closed without merging**